### DISCUSSION TITLE
MIGRATION - GCBM 3.0 (.NET 6 + .NET MAUI)

### DISCUSSION THREAD

**Discussion by laetemn** [AxionDrak/GameCube-Backup-Manager#74] [General]
I would like everyone's opinion on GCBM migration to .NET 6 + .NET MAUI platform.

I believe it would be better to release GCBM version 3.0 only when the program is completely rewritten in .NET 6 + .NET MAUI.

The two main reasons for suggesting the use of .NET 6 + .NET MAUI, account for:

- Ease of programming in C#, allowing you to reuse the current GCBM code.
- The combination of .NET 6 + .NET MAUI makes it possible to develop a multiplatform application (Android, iOS, macOS and Windows) without major difficulties.

For this, a few observations are necessary:

- The current version of GCBM would normally continue its development until it reaches version number 2.9.x.x, at which point it should be fully migrated to .NET 6 + .NET MAUI.
- No new features will be added to the current version of GCBM, only bug fixes and performance improvements should be performed.
- When migrating, GCBM will receive version number 3.0, indicating a major change in its structure and starting a new phase of the project, henceforth, multiplatform.
- The release of GCBM 3.0 could also include the implementation of a plugin system.  Making it easy to implement new features without having to directly change the main code of the program.

#### NOTE: Apparently, an application developed in .NET 6 + .NET MAUI can also run in a more organized way within the Linux operating system.

**Comment by sjohnson1021** [comment]
Already looking into this, excited to be a part of it and help any way I can.

Right now, we're stuck using C# level 7.3 ... current stable version is 11...... 8.0+ has Async streams which makes working with foreach loops asynchronously a breeze. Something I'm currently banging my head against a wall with.

There's also the huge bonus of it being cross platform. So.. Mac OS, iOS, Android, Linux, Windows .. There are some hoops to jump through when you get to dealing with OS specific code (Mainly like the system open file dialogs, etc.) but that's minor, when you can write 95% of the code once, and have it just work as a code base, then you would take the whatever steps to make that communicate with the OS, which hypothetically should only be the remaining 4.99%.

Always that .01% uncertainty clause.

Interesting resources:

https://www.youtube.com/watch?v=vTtIhUCc2sU
https://www.nuget.org/packages/CommunityToolkit.Maui
https://www.nuget.org/packages/VijayAnand.MauiTemplates/

**Comment by sjohnson1021** [comment]
As far as the UI goes, are we wanting to stay with the same layout/structure or go with like a side menu? Sort of like this: https://www.youtube.com/watch?v=N5oZnV3cA64&t=61s
<img width="1346" alt="Screenshot 2022-08-27 211056" src="https://user-images.githubusercontent.com/8071133/187054496-97aa3156-71e8-42e3-9072-c027dc387007.png">

..
Or... something completely different?

This is the Hybrid Blazor app template from the VijayAnand.MauiTemplates NuGet package:
<img width="1429" alt="Screenshot 2022-08-28 173516" src="https://user-images.githubusercontent.com/8071133/187097349-90eae614-60f8-428d-9ef1-2411ef9b9050.png">
Fluidly resizes:
<img width="370" alt="Screenshot 2022-08-28 173654" src="https://user-images.githubusercontent.com/8071133/187097376-2299a6b6-63e4-48e8-941e-945e18d7a9a0.png">
Kinda liking this one..

**Comment by sjohnson1021** [comment]
I think with the latest commit and release, barring any major overlooked or lurking bugs, the core functionality is pretty solid, and stable. Going to start stripping away TBH. Less code to sift through later. And I guess I'll mock up a few UX designs and post a poll.

Would love to hear feedback on how scanning directories (especially the destination drives) behaves now.

And feedback on this as a strongly typed hierarchy, .. needs work and I just kinda threw this together, but.. thought?: ![PSX_20220828_163333.jpg](https://user-images.githubusercontent.com/8071133/187095373-92cc8a2a-7e55-4e01-b9e5-52386502da89.jpg)

**Reply by sjohnson1021** [reply]
> I think with the latest commit and release, barring any major overlooked or lurking bugs, the core functionality is pretty solid, and stable. Going to start stripping away TBH. Less code to sift through later. And I guess I'll mock up a few UX designs and post a poll.
> 
> Would love to hear feedback on how scanning directories (especially the destination drives) behaves now.
> 
> And feedback on this as a strongly typed hierarchy, .. needs work and I just kinda threw this together, but.. thought?: ![PSX_20220828_163333.jpg](https://user-images.githubusercontent.com/8071133/187095373-92cc8a2a-7e55-4e01-b9e5-52386502da89.jpg)

With this, if we are wanting to keep a hard record of the Library, what are your thoughts on file format? JSON? XML? Something else? Could do something with like NoSQL.

Would love to get started on this as soon as possible.

**Comment by sjohnson1021** [comment]
Leaving this here. Going to go ahead and get started on... something?. at least.

https://github.com/egvijayanand/dotnet-maui-templates

```
dotnet new --install VijayAnand.MauiTemplates
dotnet new --update-check
dotnet new --update-apply
dotnet new mauiapp -n GameCubeBackupManager-MAUI -dp Hybrid
```

**Reply by laetemn** [reply]
> Deixando isso aqui. Indo em frente e começar... algo?. pelo menos.
> 
> https://github.com/egvijayanand/dotnet-maui-templates
> 
> ```
> dotnet new --install VijayAnand.MauiTemplates
> dotnet new --update-check
> dotnet new --update-apply
> dotnet new mauiapp -n GameCubeBackupManager-MAUI -dp Hybrid
> ```

I formatted the HD, installed Windows 10 and put VS2022.  Everything to have a system 100% error free and ready to work with .NET 6 and .NET MAUI.

I'm going to start doing some tests for the migration.

**Reply by sjohnson1021** [reply]
> > Deixando isso aqui. Indo em frente e começar... algo?. pelo menos.
> > https://github.com/egvijayanand/dotnet-maui-templates
> > ```
> > dotnet new --install VijayAnand.MauiTemplates
> > dotnet new --update-check
> > dotnet new --update-apply
> > dotnet new mauiapp -n GameCubeBackupManager-MAUI -dp Hybrid
> > ```
> 
> I formatted the HD, installed Windows 10 and put VS2022. Everything to have a system 100% error free and ready to work with .NET 6 and .NET MAUI.
> 
> I'm going to start doing some tests for the migration.

My current Desktop specs:
```
System Information report written at: 09/01/22 09:06:44
System Name: SEANS-DESKTOP
[System Summary]

Item	Value	
OS Name	Microsoft Windows 11 Pro	
Version	10.0.22000 Build 22000	
Other OS Description 	Not Available	
OS Manufacturer	Microsoft Corporation	
System Name	SEANS-DESKTOP	
System Manufacturer	Gigabyte Technology Co., Ltd.	
System Model	EX58-UD3R	
System Type	x64-based PC	
System SKU		
Processor	Intel(R) Core(TM) i7 CPU         920  @ 2.67GHz, 2661 Mhz, 4 Core(s), 8 Logical Processor(s)	
BIOS Version/Date	Award Software International, Inc. F11, 3/11/2010	
SMBIOS Version	2.4	
Embedded Controller Version	255.255	
BIOS Mode	Legacy	
BaseBoard Manufacturer	Gigabyte Technology Co., Ltd.	
BaseBoard Product	EX58-UD3R	
BaseBoard Version		
Platform Role	Desktop	
Secure Boot State	Unsupported	
PCR7 Configuration	Binding Not Possible	
Windows Directory	C:\Windows	
System Directory	C:\Windows\system32	
Boot Device	\Device\HarddiskVolume5	
Locale	United States	
Hardware Abstraction Layer	Version = "10.0.22000.778"	
User Name	Seans-Desktop\Sjohn	
Time Zone	Central Daylight Time	
Installed Physical Memory (RAM)	10.0 GB	
Total Physical Memory	10.0 GB	
Available Physical Memory	2.42 GB	
Total Virtual Memory	27.0 GB	
Available Virtual Memory	8.81 GB	
Page File Space	17.0 GB	
Page File	C:\pagefile.sys	
Kernel DMA Protection	Off	
Virtualization-based security	Not enabled	
Device Encryption Support	Reasons for failed automatic device encryption: TPM is not usable, PCR7 binding is not supported, Hardware Security Test Interface failed and device is not Modern Standby, Un-allowed DMA capable bus/device(s) detected, TPM is not usable	
Hyper-V - VM Monitor Mode Extensions	Yes	
Hyper-V - Second Level Address Translation Extensions	Yes	
Hyper-V - Virtualization Enabled in Firmware	Yes	
Hyper-V - Data Execution Protection	Yes
```

I've also got a Laptop in a Triple-boot set-up running Windows 11, Linux Mint 20, and Mac OS (I forget which one, its a Hackintosh build).

```
HP Probook 4540s
1tb HD
500gb SSD
8gb Ram
Intel i9 (but like.. really Old i9)
```

**Comment by sjohnson1021** [comment]
> I would like everyone's opinion on GCBM migration to .NET 6 + .NET MAUI platform.
> 
> I believe it would be better to release GCBM version 3.0 only when the program is completely rewritten in .NET 6 + .NET MAUI.
> 
> The two main reasons for suggesting the use of .NET 6 + .NET MAUI, account for:
> 
> * Ease of programming in C#, allowing you to reuse the current GCBM code.
> * The combination of .NET 6 + .NET MAUI makes it possible to develop a multiplatform application (Android, iOS, macOS and Windows) without major difficulties.
> 
> For this, a few observations are necessary:
> 
> * The current version of GCBM would normally continue its development until it reaches version number 2.9.x.x, at which point it should be fully migrated to .NET 6 + .NET MAUI.
> * No new features will be added to the current version of GCBM, only bug fixes and performance improvements should be performed.
> * When migrating, GCBM will receive version number 3.0, indicating a major change in its structure and starting a new phase of the project, henceforth, multiplatform.
> * The release of GCBM 3.0 could also include the implementation of a plugin system.  Making it easy to implement new features without having to directly change the main code of the program.
> 
> #### NOTE: Apparently, an application developed in .NET 6 + .NET MAUI can also run in a more organized way within the Linux operating system.

# Hey, dunno how far or how much time you've had to invest in a local development branch, but I figured I'd give you a heads up on some progress I've made. 

### I've fully isolated retrieval of game information to Game.cs with the latest push to main.
_(If you're brave, check out my experimental branch, that's still winforms but.. some fun stuff happening there.)_ 


You can use the Game class as a utility class by Initiating an empty game.

`Game gameUtilities = new Game()`

and then use this globally for Game based logic (getting titles, publisher, internal name, etc) by calling the instance variable

`Game MarioKart = gameUtilities.GetGameInfo("C:\Roms\MarioKart.iso")`

this will return a fully populated Game object which you can then access normally.

`Text = MarioKart.Title`
`ID = MarioKart.ID`
👍 😄

**Comment by sjohnson1021** [comment]
@laetemn Why do we have GenerateTreeView?

Was this for development purposes? or does this actually verify file integrity?

I've removed it locally in my development branch.. and... it still works? I want to make sure I can scrap this, before well.. scrapping it. Otherwise I can find a way to .. rewrite it? TreeNode is Winforms only, and unless you know of an alternative? .. Though this would be mainly for visual as it is .. win FORMS.. so the functionality can be replicated with a more basic data structure, which would improve performance. But upon debugging and adding breakpoints, It hits only in ReadImageTOC and ReadImageDiscTOC, commenting it out. No explosions happen, and the program behaves as if nothing has changed.?

If I can safely scrap this, let me know please?

**Reply by laetemn** [reply]
> @laetemn Why do we have GenerateTreeView?
> 
> Was this for development purposes? or does this actually verify file integrity?
> 
> I've removed it locally in my development branch.. and... it still works? I want to make sure I can scrap this, before well.. scrapping it. Otherwise I can find a way to .. rewrite it? TreeNode is Winforms only, and unless you know of an alternative? .. Though this would be mainly for visual as it is .. win FORMS.. so the functionality can be replicated with a more basic data structure, which would improve performance. But upon debugging and adding breakpoints, It hits only in ReadImageTOC and ReadImageDiscTOC, commenting it out. No explosions happen, and the program behaves as if nothing has changed.?
> 
> If I can safely scrap this, let me know please?

Some things in the code are duplicated, but in a different way. Basically for performance testing.
They can be removed (refactored) without any problems.

**Reply by sjohnson1021** [reply]
> > @laetemn Why do we have GenerateTreeView?
> > Was this for development purposes? or does this actually verify file integrity?
> > I've removed it locally in my development branch.. and... it still works? I want to make sure I can scrap this, before well.. scrapping it. Otherwise I can find a way to .. rewrite it? TreeNode is Winforms only, and unless you know of an alternative? .. Though this would be mainly for visual as it is .. win FORMS.. so the functionality can be replicated with a more basic data structure, which would improve performance. But upon debugging and adding breakpoints, It hits only in ReadImageTOC and ReadImageDiscTOC, commenting it out. No explosions happen, and the program behaves as if nothing has changed.?
> > If I can safely scrap this, let me know please?
> 
> Some things in the code are duplicated, but in a different way. Basically for performance testing. They can be removed (refactored) without any problems.

Cool cool, so on that note.... Do you mind/ are you available to test a build? either I can Push to a branch, or compile a build? It's .. solid, and I'm happy with it. I just don't want to get excited and push 2.6.1 before it's fully ready. Batch scrubbing is fully functional, interface is more responsive. and I've added an option to disable notifications.

https://user-images.githubusercontent.com/8071133/188525662-dfeb3407-b3db-4631-841e-019ed50f443b.mp4

**Comment by sjohnson1021** [comment]
https://stackoverflow.com/questions/72780912/net-blazor-desktop-folder-picker

... Really having a hard time figuring out which direction to go... I'm not really sure a Blazor Hybrid app is the way, .. but I also don't know that it isn't.... It's very much a newborn language that is still being built, and I'm not sure what the limitations are or will be. I'm not fully sure that Blazor will be appropriate or even capable of the kind of file access required.. 

Then I run into snippets Like this and the link above that give me hope.

```
    string fileName = Path.Combine(
        Environment.GetFolderPath(
            Environment.SpecialFolder.LocalApplicationData), "temp.csv");
```

**Comment by sjohnson1021** [comment]
@laetemn Alright. Think I've done just about all of the optimization I can manage without learning and switching to C++ or something lol. Full steam ahead on Maui.. I know you just dropped 2.6.0.10,... But. uhm.. I really wish you had waited a day 😛. If you want to build the latest commit, or just take a peek at the scrub code, it might give you an idea. I don't think it actually scrubs the games any faster, that's gcit's limitation, but the interface stays awake and progress is handled.. well. better. 
https://youtu.be/DgMb9tM_JfU

New release is out, not sure why auto updater isnt recognizing .. it should be all hooked up?

**Comment by sjohnson1021** [comment]
# [File System Helpers](https://docs.microsoft.com/en-us/dotnet/maui/platform-integration/storage/file-system-helpers)

## Cache directory
Cache directory
> > To get the application's directory to store cache data. Cache data can be used for any data that needs to persist longer than temporary data, but shouldn't be data that is required to operate the app, as the operating system may clear this storage.
`string cacheDir = FileSystem.Current.CacheDirectory;`

## App data directory
> > To get the app's top-level directory for any files that aren't user data files. These files are backed up with the operating system syncing framework.
`string mainDir = FileSystem.Current.AppDataDirectory;`

## [Pick a file](https://docs.microsoft.com/en-us/dotnet/maui/platform-integration/storage/file-picker?tabs=android) 
> > The PickAsync method prompts the user to pick a file from the device. Use the PickOptions type to specify the title and file types allowed with the picker. The following example demonstrates opening the picker and processing the selected image:
```
public async Task<FileResult> PickAndShow(PickOptions options)
{
    try
    {
        var result = await FilePicker.Default.PickAsync(options);
        if (result != null)
        {
            if (result.FileName.EndsWith("jpg", StringComparison.OrdinalIgnoreCase) ||
                result.FileName.EndsWith("png", StringComparison.OrdinalIgnoreCase))
            {
                using var stream = await result.OpenReadAsync();
                var image = ImageSource.FromStream(() => stream);
            }
        }

        return result;
    }
    catch (Exception ex)
    {
        // The user canceled or something went wrong
    }

    return null;
}
```

> > Default file types are provided with FilePickerFileType.Images, FilePickerFileType.Png, and FilePickerFilerType.Videos. You can specify custom file types per platform, by creating an instance of the FilePickerFileType class. The constructor of this class takes a dictionary that is keyed by the DevicePlatform type to identify the platform. The value of the dictionary key is a collection of strings representing the file types. For example here's how you would specify specific comic file types:

```
var customFileType = new FilePickerFileType(
                new Dictionary<DevicePlatform, IEnumerable<string>>
                {
                    { DevicePlatform.iOS, new[] { "public.my.comic.extension" } }, // or general UTType values
                    { DevicePlatform.Android, new[] { "application/comics" } },
                    { DevicePlatform.WinUI, new[] { ".cbr", ".cbz" } },
                    { DevicePlatform.Tizen, new[] { "*/*" } },
                    { DevicePlatform.macOS, new[] { "cbr", "cbz" } }, // or general UTType values
                });

PickOptions options = new()
{
    PickerTitle = "Please select a comic file",
    FileTypes = customFileType,
};
```

##  Pick multiple files
> > If you want the user to pick multiple files, call the FilePicker.PickMultipleAsync method. This method also takes a PickOptions parameter to specify additional information. The results are the same as PickAsync, but instead of the FileResult type returned, an IEnumerable<FileResult> type is returned with all of the selected files.
> > > The FullPath property doesn't always return the physical path to the file. To get the file, use the OpenReadAsync method.

https://stackoverflow.com/q/70999773/19660393
